internal/resources/cluster/data_source_cluster.go

```diff
@@ -129,18 +129,53 @@ func dataSourceTMCClusterRead(ctx context.Context, d *schema.ResourceData, m int
 
 	clusterSpec := constructSpec(d)
 
-	switch {
-	case resp.Cluster.Spec.TkgAws != nil:
-		resp.Cluster.Spec.TkgAws.Topology.NodePools = clusterSpec.TkgAws.Topology.NodePools
-	case resp.Cluster.Spec.TkgServiceVsphere != nil:
-		resp.Cluster.Spec.TkgServiceVsphere.Topology.NodePools = clusterSpec.TkgServiceVsphere.Topology.NodePools
-	case resp.Cluster.Spec.TkgVsphere != nil:
-		resp.Cluster.Spec.TkgVsphere.Topology.NodePools = clusterSpec.TkgVsphere.Topology.NodePools
-	}
+	getNodepoolForCluster(resp.Cluster.Spec, clusterSpec)
 
 	if err := d.Set(SpecKey, flattenSpec(resp.Cluster.Spec)); err != nil {
 		return diag.FromErr(err)
 	}
 
 	return diags
 }
+
+func getNodepoolForCluster(respSpec, clusterSpec *clustermodel.VmwareTanzuManageV1alpha1ClusterSpec) {
+	if clusterSpec == nil {
+		return
+	}
+
+	switch {
+	// need the check for data-source read where the spec is not populated.
+	case clusterSpec.TkgAws == nil && clusterSpec.TkgServiceVsphere == nil && clusterSpec.TkgVsphere == nil:
+		break
+	case clusterSpec.TkgAws != nil && clusterSpec.TkgAws.Topology != nil:
+		if respSpec.TkgAws == nil {
```

Comment:
This is code is wrong.

The intention of the read function is reflect the latest state of the server. Here you are reverting to terraform state which is not expected.

This function should be only scoped to node pools updated from state to avoid diff.

Reply:
Made changes as per the discussion and tested.